app/Http/Controllers/Auth/SocialiteController.php

```diff
@@ -164,7 +164,13 @@ private function setRolesFromClaim(string $provider, $user): bool
             ! empty($claims)
         ) {
             $roles = [];
-            $attributes = $this->socialite_user->getRaw();
+            $attributes = [];
+            $attribute_list = $this->socialite_user->getRaw();
+            foreach ($attribute_list as $a) {
+                $attribute_name = $a->getName();
```

Comment:
i dont not see how this works, first of all `$a` should be just a string or so in this case? and `getName()` has the comment `Get the full name of the user.`
Have you tested this? 
Have you tested it with OAuth2 also?

Reply:
- output of `error_log(print_r($attribute_list, TRUE), 3 , 'errorlog.log');` below.
- I have tested it with SAML and it works fine.
- I haven't tested with OAuth2, though I can't see how the original code could ever have worked there either.


```
(
    [0] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => nickname
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => Joe
                )

        )

    [1] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => preferred_username
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => joeb
                )

        )

    [2] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => fullname
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => Joe Blogs
                )

        )

    [3] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => email
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => joeb@email.example.com
                )

        )

    [4] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => givenname
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => Joe
                )

        )

    [5] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => surname
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => Blogs
                )

        )

    [6] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => username
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => joeb
                )

        )

    [7] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => gecos
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => Joe Blogs
                )

        )

    [8] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => mail
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => joeb@email.example.com
                )

        )

    [9] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => sn
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => Blogs
                )

        )

    [10] => LightSaml\Model\Assertion\Attribute Object
        (
            [name:protected] => groups
            [nameFormat:protected] => urn:oasis:names:tc:SAML:2.0:attrname-format:basic
            [friendlyName:protected] =>
            [attributeValue:protected] => Array
                (
                    [0] => librenms_admins
                    [1] => other_group
                )

        )

)
```

Reply:
If I add this to Okta authed it explodes

~/logs/librenms.log
`Call to a member function getName() on string {"userId":15,"exception":"[object] (Error(code: 0): Call to a member function getName() on string at /opt/librenms/app/Http/Controllers/Auth/SocialiteController.php:169)"}
`
cat ~/html/errorlog.log
```
Array
(
    [sub] => 00u1qi5ae4888hoc70h8
    [name] => REDACTED
    [locale] => en_US
    [email] => REDACTED
    [preferred_username] => REDACTED
    [given_name] => Joe
    [family_name] => Smith
    [zoneinfo] => America/Los_Angeles
    [updated_at] => 1715015601
    [email_verified] => 1
    [groups] => Array
        (
            [0] => LibreNMSNonProd-Write
        )

)

```

Reply:
my get_class($this->socialite_user) is SocialiteProviders\Manager\OAuth2\User ... you might be able to gate your changes around whatever your $this->socialite_user is an instanceOf ?

Reply:
I did as you suggested (check for instanceof \SocialiteProviders\Saml2\User) and it seems to work.
Now I'm being led down another rabbit hole with failed linting.
Since PHP & git ain't my expertise, I'm giving up on this.
Thanks for your input nonetheless.

Reply:
Yeah, phptan can't find \SocialiteProviders\Saml2\User.  What PHP package provides this class?